rdkit/ML/Scoring/Scoring.py

```diff
@@ -14,6 +14,40 @@
 """
 
 import math
+import random  # Does rdkit depend on numpy? => Use np.random instead
+import copy
+from operator import itemgetter
+
+
+def avoid_sorting_artifacts_on_ranking(scores, col, reverse=False):
+  """
+  Safeguard against rotten ROCs (& AUCs & RIEs and whatnot) due to ordering artifacts.
+
+  Avoids having artificially good/bad scores because:
+    - `scores` are ordered by class (col) and python stable (tim)sort
+      keeps the order on ties; this is specially bad if our model
+      returns garbage scores, let them be nans or any other constant.
+    - related, we leak the class in sorting
+
+  These problems are not uncommon. See the unit tests for some artificial examples.
+  """
+  # Shuffle
+  scores = copy.copy(scores)        # No side effects
```

Comment:
`copy.copy` is shallow, and allows objects to define their own copy mechanism, which is used, for example, by numpy - curiously numpy does copy its data buffer, which is correct behavior here.

However, you make me wonder, and I think you have a very good point here. I need to do some proper benchmarking but I think here we are facing a speed/memory tradeoff. Think about scores being a numpy array. Using `scores = list(scores)` will generate a list of views pointing to the original array (if scores is large this will be a lot of python objects and make memory explosion an issue). However, shuffling will just be much faster on the list (moving around pointers) than on the array (swapping rows in the array). Then later in the function we use "sorted" which will anyway generate a list so I'm inclined to use your suggestion and just use a list. The alternative would be to split the code in two branches depending on whether `scores` is an array or not, which would make the function way more complicated.

In any case, anyone wanting high performance AUC computations should look somewhere else. For example, when I need performance, I use a [cython implementation](https://github.com/sdvillal/ccl-malaria/blob/f7acf0535b50080797e4844e3c2ebf50f3f1b92b/src/minioscail/common/faster.pyx) (that does handle ties more efficiently by averaging rankings). I only need performace when optimizing some parameters based on AUC over large datasets, but in 99% of the cases performance of AUC computation is just irrelevant.